lib/server.js

```diff
@@ -243,8 +378,205 @@ function createServer(options, clients) {
+                        req.headers.Authorization ||
+                        '';
+
+        var authHeaderLower = authHeader.toLowerCase();
+        var startsWithAws4 = authHeaderLower.indexOf('aws4-hmac-sha256') === 0;
+        var host = req.headers.host || '';
+
+        // Simple and clear S3 detection: ONLY SigV4 requests are S3 requests
+        // Traditional Manta requests use HTTP Signature auth, not SigV4
+        // Regardless of host format (virtual-hosted or path-style),
+        // only SigV4 = S3
+        var isSigV4 = startsWithAws4;
+        var isS3Format = isSigV4;
+
+        req.log.debug({
+            method: req.method,
+            path: req.path(),
+            host: host,
+            isSigV4: isSigV4,
+            isS3Format: isS3Format,
+            authHeaderFirst50: authHeader.substring(0, 50),
+            detectionReason: isSigV4 ? 'SigV4_AUTH' : 'NOT_S3'
+        }, 'S3_DEBUG: Strict S3 detection - Only SigV4 = S3');
+
+        if (isS3Format) {
+            req.log.debug('S3_DEBUG: DETECTED S3 REQUEST');
+            req.isS3Request = true;
+            req.s3Request = s3Compat.parseS3Request(req);
+            req.log.debug({
+                s3Request: req.s3Request
+            }, 'S3_DEBUG: Parsed S3 request details');
+
+            // For S3 object operations (uploads AND downloads),
+            // mark to skip S3 response processing
+            if ((req.method === 'PUT' &&
+                req.s3Request.operation === 'CreateBucketObject') ||
+                (req.method === 'GET' &&
+                  req.s3Request.operation === 'GetBucketObject')) {
+                req.log.debug('S3_DEBUG: Marking S3 object operation'+
+                ' to skip response processing (method=' + req.method + ')');
+
+                // Mark this request to skip ALL S3 response processing
+                req._skipS3ResponseProcessing = true;
+                req._binaryOperation = true;
+            }
+        } else {
+           // req.log.debug('S3_DEBUG: Not S3 request'+
+           // ' - using traditional Manta routing');
+            req.isS3Request = false;
+        }
+
+        next();
+    });
+
+    // NOTE: S3 root handler will be added after authentication middleware
+
+    // Set up other routes (ping, docs, crossdomain)
+    server.get({
+        path: '/ping',
+        name: 'ping'
+    }, function ping(req, res, next) {
+        res.set('Connection', 'close');
+        res.set('Content-Length', 0);
+        res.writeHead(200);
+        res.end();
+        next();
+    });
+
+    server.get(/^\/docs\/?/, function redirect(req, res, next) {
+        res.set('Content-Length', 0);
+        res.set('Connection', 'keep-alive');
+        res.set('Date', new Date());
+        res.set('Location', 'http://apidocs.joyent.com/manta/');
+        res.send(302);
+        next(false);
+    });
+
+    server.get('/crossdomain.xml', function flashXML(req, res, next) {
```

Comment:
WTF is Adobe Flash doing here?!?!?!?

crossdomain in general is bad, AFAIC, but Adobe?!?

If I'm missing something please enlighten me, but this looks bad at first glance.